content/blog/2025-12-14-lel-final-project.md

```diff
@@ -0,0 +1,167 @@
+| `latex-formulas`  |       552,339 |          506,900 |          360 |          119  |
+| arXiv|        42,896 |           29,356 |           33 |            9  |
+### Speed
+For functions that make it through parsing and reference evaluation, I also measure performance by timing the entire pipeline, from normalization and parsing through LLVM codegen, clang compilation, and execution of the test binary. On a M4 MacBook Pro laptop with 24 GB RAM running the Sequoia 15.6.1, processing 345 such functions takes 18.851 seconds total, for an average of about 54.6 ms per function, with a median of 0.1 ms and a maximum of ~512 ms. 
+
+### On the tiny hand-written test
+On the curated suite of 24 hand-written LEL functions covering arithmetic, powers, fractions, zero-argument functions, conditional expressions, and calls, our pipeline successfully parses, compiles, and validates 22/22 (100%) non-recursive functions. The two recursive examples (fact, accum) are intentionally treated as unsupported by our evaluation harness to avoid non-terminating test runs.
+
+### On the Hugging face latex-formulas dataset 
+On the large latex-formulas corpus (HuggingFace), 506,900 (about 92%) are successfully normalized by MiTeX into Typst math. 360 matched the LEL parser and are evaluated. Out of these, 119 (33%) functions passed, 226 failed with runtime_error, and 15 were marked unsupported due to explicit recursion (e.g., f(a) = f(alpha) + …, T(x) = T(-x)). The runtime_error cases are almost entirely due to my reference evaluator’s strict handling of free symbols: many equations involve undeclared constants, matrix/vector symbols, or built-in functions (e.g., x = f(x)), and the evaluator reports these as unknown variable rather than assuming numeric values. In contrast, the 119 successful cases are the clean subset where all identifiers can be interpreted purely as scalar parameters and local variables under my simple semantics. 
+
+### On the ArXiv dataset
+On the arXiv robotics corpus, MiTeX successfully produced normalized Typst math for 29,356 equations. Out of these, only 33 equations (~0.1%) matched the very restrictive LEL shape and were accepted by the LEL parser. When I ran the end-to-end evaluation pipeline on these 33 functions, 9 passed (27%) and 24 were reported as runtime_error. The failures were not due to codegen bugs but to the reference evaluator’s deliberately conservative treatment of free symbols: many equations use undeclared constants or matrix-valued symbols (e.g., Ax = b, R = det(M) - k(trace(M))^2), which the evaluator flags as "unknown variable" rather than silently treating them as numeric parameters. On real papers, the LEL compiler can successfully compile and run a small handful of clean scalar equations, while most extracted equations either fall outside the LEL subset or rely on ambient symbols (vectors, matrices, constants) that are not modeled in my current semantics. 
```

Comment:
Consider using math mode or just `code backticks` to make these bits look less weird in the final rendered page:

> Ax = b, R = det(M) - k(trace(M))^2